### ISSUE TITLE
ShellCheck: differentiate POSIX and Bash

### ISSUE THREAD

**Issue by skovhus** [bash-lsp/bash-language-server#576]
From https://github.com/bash-lsp/bash-language-server/issues/104#issuecomment-1327720330 by @Jackenmen

It seems that this language server doesn't show all messages that `shellcheck` does. It seems related to not considering POSIX sh and Bash different.

Here's the output from bash-language-server:
```
/tmp/blah.sh:
    3:10  warning SC2051: Bash doesn't support variables in brace range expansions. ​shellcheck:2051
   10:8   hint    SC2007: Use $((..)) instead of deprecated $[..] ​shellcheck:2007
   12:10  warning SC2154: strbuzz is referenced but not assigned. ​shellcheck:2154
```
Here's the output from diagnostic-languageserver running `shellcheck`:
```
/tmp/blah.sh:
    3:10  warning In POSIX sh, brace expansion is undefined. [3009] ​shellcheck
    3:14  warning In POSIX sh, RANDOM is undefined. [3028] ​shellcheck
    6:6   warning In POSIX sh, standalone ((..)) is undefined. [3006] ​shellcheck
   10:8   warning In POSIX sh, $[..] in place of $((..)) is undefined. [3007] ​shellcheck
   10:8   hint    Use $((..)) instead of deprecated $[..] [2007] ​shellcheck
   10:15  warning In POSIX sh, == in place of = is undefined. [3014] ​shellcheck
   12:10  warning strbuzz is referenced but not assigned. [2154] ​shellcheck
   14:6   warning In POSIX sh, [[ ]] is undefined. [3010] ​shellcheck
```
Here's the raw output from `shellcheck`:
```
In blah.sh line 3:
for n in {1..$RANDOM}
         ^----------^ SC3009 (warning): In POSIX sh, brace expansion is undefined.
             ^-----^ SC3028 (warning): In POSIX sh, RANDOM is undefined.


In blah.sh line 6:
  if (( n % 3 == 0 ))
     ^--------------^ SC3006 (warning): In POSIX sh, standalone ((..)) is undefined.


In blah.sh line 10:
  if [ $[n%5] == 0 ]
       ^----^ SC3007 (warning): In POSIX sh, $[..] in place of $((..)) is undefined.
       ^----^ SC2007 (style): Use $((..)) instead of deprecated $[..]
              ^-- SC3014 (warning): In POSIX sh, == in place of = is undefined.


In blah.sh line 12:
    str="$strbuzz"
         ^------^ SC2154 (warning): strbuzz is referenced but not assigned.


In blah.sh line 14:
  if [[ ! $str ]]
     ^----------^ SC3010 (warning): In POSIX sh, [[ ]] is undefined.

For more information:
  https://www.shellcheck.net/wiki/SC2154 -- strbuzz is referenced but not ass...
  https://www.shellcheck.net/wiki/SC3006 -- In POSIX sh, standalone ((..)) is...
  https://www.shellcheck.net/wiki/SC3007 -- In POSIX sh, $[..] in place of $(...
```

And the example file:
```sh
#!/bin/sh

for n in {1..$RANDOM}
do
  str=""
  if (( n % 3 == 0 ))
  then
    str="fizz"
  fi
  if [ $[n%5] == 0 ]
  then
    str="$strbuzz"
  fi
  if [[ ! $str ]]
  then
    str="$n"
  fi
  echo "$str"
done
```

**Comment by skovhus** [comment]
This is now supported based on the shebang in `bash-language-server@4.0.0-beta.1` (that now requires node >= 14, hence the new major version bump and the beta tag while reiterating several features). This has also been released as the vscode extension 1.18.0.

@Jackenmen can I get you to test this out?

I'm wondering if we should also use the file extension (in case there isn't a shebang), what do you think?

**Comment by skovhus** [comment]
To install the beta release you can run `npm i -g bash-language-server@4.0.0-beta.1`

**Comment by Shane-XB-Qian** [comment]
1, looks not recognized `#!/bin/env sh` style shebang.
2, nowdays looks most users perhaps assumed bash is the end version, i mean some sh script marked as `/bin/sh` but seems actually they thought that is 'bash', now we had to modify those shebang to `/bin/bash`?

**Comment by skovhus** [comment]
> 1, looks not recognized #!/bin/env sh style shebang.

I haven't seen that shebang before, but it should be supported in the next version.

> 2, nowdays looks most users perhaps assumed bash is the end version, i mean some sh script marked as /bin/sh but seems actually they thought that is 'bash', now we had to modify those shebang to /bin/bash?

Seems like a good thing we help remind people that there is a difference here.

**Comment by Jackenmen** [comment]
Quite a bit late since you already released with stable version 4.1.0 but I did just test it out and it works great, thanks so much for working on this!
